lsbi/stats.py

```diff
@@ -1,165 +1,194 @@
 
         Returns
         -------
-        marginalised distribution: multivariate_normal
+        logpdf : array_like
+            Log of the probability density function evaluated at x.
+            if not broadcast: shape `(*size, *self.shape)`
+            else: shape broadcast of `(*size,) and `self.shape`
         """
-        i = self._bar(indices)
-        mean = self.mean[i]
-        cov = self.cov[i][:, i]
-        return multivariate_normal(mean, cov)
+        x = np.array(x)
+        if broadcast:
+            dx = x - self.mean
+        else:
+            size = x.shape[:-1]
+            mean = np.broadcast_to(self.mean, (*self.shape, self.dim))
+            dx = x.reshape(*size, *np.ones_like(self.shape), self.dim) - mean
+        if self.diagonal:
+            chi2 = (dx**2 / self.cov).sum(axis=-1)
+            norm = -np.log(2 * np.pi * np.ones(self.dim) * self.cov).sum(axis=-1) / 2
+        else:
+            chi2 = np.einsum("...j,...jk,...k->...", dx, inv(self.cov), dx)
+            norm = -logdet(2 * np.pi * self.cov) / 2
+        return norm - chi2 / 2
 
-    def condition(self, indices, values):
-        """Condition on indices with values.
+    def pdf(self, x, broadcast=False):
+        """Probability density function.
 
         Parameters
         ----------
-        indices : array_like
-            Indices to condition over.
-        values : array_like
-            Values to condition on.
+        x : array_like, shape `(*size, dim)`
+            Points at which to evaluate the probability density function.
+        broadcast : bool, optional, default=False
+            If True, broadcast x across the distribution parameters.
 
         Returns
         -------
-        conditional distribution: multivariate_normal
+        pdf : array_like
+            Probability density function evaluated at x.
+            if not broadcast: shape `(*size, *self.shape)`
+            else: shape broadcast of `(*size,) and `self.shape`
         """
-        i = self._bar(indices)
-        k = indices
-        mean = self.mean[i] + self.cov[i][:, k] @ inv(self.cov[k][:, k]) @ (
-            values - self.mean[k]
-        )
-        cov = (
-            self.cov[i][:, i]
-            - self.cov[i][:, k] @ inv(self.cov[k][:, k]) @ self.cov[k][:, i]
-        )
-        return multivariate_normal(mean, cov)
-
-    def _bar(self, indices):
-        """Return the indices not in the given indices."""
-        k = np.ones(len(self.mean), dtype=bool)
-        k[indices] = False
-        return k
-
-    def bijector(self, x, inverse=False):
-        """Bijector between U([0, 1])^d and the distribution.
-
-        - x in [0, 1]^d is the hypercube space.
-        - theta in R^d is the physical space.
+        return np.exp(self.logpdf(x, broadcast=broadcast))
 
-        Computes the transformation from x to theta or theta to x depending on
-        the value of inverse.
+    def rvs(self, size=()):
```

Comment:
lsbi multivariate normal differs from scipy in not squeezing for single rvs

```python
scipy.stats.multivariate_normal(np.ones(100)).rvs(1).shape
(100,)
lsbi.stats.multivariate_normal(np.ones(100)).rvs(1).shape
(1, 100)
```

Reply:
This was a conscious choice, mostly because it makes the code & tests considerably easier to write systematically. Is it a dealbreaker?

Reply:
Nope, no issue from my PoV just thought I would flag it as a potential divergence from base structure

Reply:
The default equivalence of rvs(1) and rvs() is arguably a little perverse
```
scipy.stats.multivariate_normal(np.ones(100)).rvs().shape
(100,)
lsbi.stats.multivariate_normal(np.ones(100)).rvs().shape
(100,)
```